README.md

```diff
@@ -113,6 +105,73 @@ return {
 }
 ```
 
+## Recipes
+
+<details><summary>Filter commands by tag</summary>
+
+### Configuration
+
+```lua
+opts = {
+  commands = {
+    { name = "Copy full path", execute = ":let @+ = expand('%:p')", tags = { "clipboard" } },
+    { name = "Copy filename", execute = ":let @+ = expand('%:t')", tags = { "clipboard" } },
+    { name = "Remove empty lines", execute = ":g/^$/d", tags = { "editor", "newline" } },
+  },
+}
+```
+
+### Usage
+
+```lua
+-- Show only clipboard commands
+local function filter(command)
+  return command.tags ~= nil and vim.tbl_contains(command.tags, "clipboard")
+end
+require("toolbox").show_picker(filter, { prompt = "Select clipboard command" })
+```
+
+</details>
+
+<details><summary>Filter commands by current filetype</summary>
+
+### Configuration
+
+```lua
+opts = {
+  commands = {
+    { name = "Copy full path", execute = ":let @+ = expand('%:p')" },
+    { name = "Format JSON with jq", execute = ":%!jq", filetype = "json" }
+    { name = "Format QML file", execute = ":qmlformat %", filetype = "qml" }
+  },
+}
+```
+
+### Usage
+
+```lua
+require("toolbox").show_picker(function(command)
+  return command.filetype == vim.bo.filetype
+end, { prompt = "Select " .. vim.bo.filetype .. " command" })
+```
+
+</details>
+
+<details><summary>Change command representation in the list</summary>
+
+### Usage
+
+```lua
+require("toolbox").show_picker(nil, {
+  format_item = function(command)
+    -- Display => and execute string after the name
+    return command.name .. " => " .. (type(command.execute) == "function" and "<function>" or command.execute)
+  end
+})
+```
+
+</details>
+
```

Comment:
10 last commands feature needs extra functionality.
First of all, I need to introduce callback:
```lua
opts = {
  commands = {},
  pre_run = function(command)
    -- save timestamp for the executed command to some table
  end,
  post_run = function(command, success)
    -- added for consistency with pre_run
    -- potential use case - notify about the executed command name, success or failure (info or warning)
  end
}
```

After this, I should save usage timestamps in data, to be available in the next session

Let's do this in another PR

Reply:
hmm, this is a good point, are there any other scenarios that do not require this and the 'tags' implementation in main cannot support?

Reply:
The problem with tags is that the user should manually set them, and tags are a limited set of values. With tags, the user can't filter by integers, double values, or dynamically calculated values (like the last usage timestamp).

More examples that can't be covered with tags:

1. Display commands that were never used
2. Display the ten most popular commands
3. Display commands that were applied to the current project (or the current directory) earlier
And more ...

Note that examples 1-3 are similar to the "10 recent" case and need `pre_run` callback and storing session data

Reply:
Let me think if I want to land this without pre_run and post_run, since all the listed scenarios that the change covers require that.

Also they probably should be attached to `show_picker` instead of opts, however the presence of require_input also complicates thing abit.

Reply:
Examples of filters without the `pre_run`:
- Show commands that start with "Copy" in the name.
- Show commands with Lua functions only.
- Show commands with the word "Format".
- Show commands with 1 or more arguments.

In all cases, it's easier to write a universal function than set tags and maintain them.

Reply:
If you'd like, we can land the vim ui select first in a different PR. I am still on the fence a bit for this

Reply:
I'm OK to merge select_opts first

Reply:
Lets merge select_opts first. I am very thankful for you spending your time on this but give me a few days to think through this.

Reply:
Please remember that if the plugin is oriented on custom properties, it can be used in filtering and sorting. Tags are unusable for sorting

Reply:
I will think about it. Thank you.

Reply:
Hi @fpohtmeh, I've given it a thought. I would like to request to do this enhancement in a new function, perhaps `show_custom_picker()` or `show_advanced_picker()` so that for advanced users that want advanced filtering and or sorting in the future, they can use that, while we remain the current behaviour for the default show_picker.

I can help implement it, the default show_picker() will just call show_custom_picker under the hood.

What do you think of this idea?

Reply:
![image](https://github.com/user-attachments/assets/6f64b8f9-8ff8-4053-9c53-135fca21a78b)

Reply:
If you are ok with this, I will push the changes, update the documentation and proceed to merge it.

Reply:
Yes, please proceed

Reply:
But my opinion about the tag is the same - it covers some of your specific use cases.
Users lose nothing if you remove tags but introduce a filter and a recipe with a tag. And the plugin API will be simpler